java/AsyncIO/src/main/java/com/amazonaws/services/msf/ProcessingFunction.java

```diff
@@ -0,0 +1,86 @@
+package com.amazonaws.services.msf;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.streaming.api.functions.async.ResultFuture;
+import org.apache.flink.streaming.api.functions.async.RichAsyncFunction;
+import org.asynchttpclient.AsyncHttpClient;
+import org.asynchttpclient.DefaultAsyncHttpClientConfig;
+import org.asynchttpclient.Dsl;
+import org.asynchttpclient.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.Properties;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+
+public class ProcessingFunction extends RichAsyncFunction<IncomingEvent, ProcessedEvent> {
+
+    public Properties endpointProperties;
+    private transient AsyncHttpClient client;
+
+    ProcessingFunction(Properties endpointProperties) {
+        this.endpointProperties = endpointProperties;
+
+    }
+
+    private static final Logger LOG = LoggerFactory.getLogger(ProcessingFunction.class);
+
+    /**
+     * Instantiate the connection to an async client here to use within asyncInvoke
+     */
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        DefaultAsyncHttpClientConfig.Builder clientBuilder = Dsl.config().setConnectTimeout(Duration.ofSeconds(10));
+        client = Dsl.asyncHttpClient(clientBuilder);
+
+    }
+
+    @Override
+    public void asyncInvoke(IncomingEvent incomingEvent, ResultFuture<ProcessedEvent> resultFuture) {
+
+        // Create a new ProcessedEvent instance
+        ProcessedEvent processedEvent = new ProcessedEvent(
+                incomingEvent.getMessage(),
+                null);
+        LOG.debug("new request {}", incomingEvent);
+
+        // Defining API request
+        Future<Response> future = client.prepareGet(endpointProperties.getProperty("api.url"))
```

Comment:
@nicusX do you have an example that does this? I want to make sure I'm doing it in the same way.

Reply:
We don't have many examples of custom operators with parameters in this repo. I think the closest example is this [RichMapFunction](https://github.com/aws-samples/amazon-managed-service-for-apache-flink-examples/blob/540208e2cc50da191904632ef12d9bd30c605a68/java/CustomMetrics/src/main/java/com/amazonaws/services/msf/MetricEmittingMapperFunction.java#L18): the idea is simply passing to the mapFunction parameters like API URL as typed parameters an keeping them in (final) typed properties of the object. 
Alternatively, you can validate the `Properties` in the constructor of the MapFunction, using `Preconditions`. The result would be the same: the job will fail to initialize, instead of failing and restarting because the operator fails initializing.

I slightly prefer typed parameters to generic `Properties`, for clarity. Unless you need to pass *whatever properties* to an underlying client, as it happens with Kafka or Kinesis connectors. But it's a matter of preferences. What is important is that the job failing to initialize when a parameter is missing is much more user friendly

Reply:
you can check the most recent changes for this, using the Preconditions library--job will not start if any necessary properties are missing.